pkg/hub/controllers/finalizercontroller/add_finalizer_controller.go

```diff
@@ -0,0 +1,158 @@
+package finalizercontroller
+
+import (
+	"context"
+	"strings"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	rbacv1informers "k8s.io/client-go/informers/rbac/v1"
+	rbacv1client "k8s.io/client-go/kubernetes/typed/rbac/v1"
+	rbacv1listers "k8s.io/client-go/listers/rbac/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+const (
+	manifestWorkFinalizer = "cluster.open-cluster-management.io/manifest-work-cleanup"
+	keySuffix             = ":spoke-work"
+)
+
+// AddFinalizerController is to add the cluster.open-cluster-management.io/manifest-work-cleanup
+// finalizer on role/rolebinding for work agent in spoke cluster namespace.
+type AddFinalizerController struct {
+	roleLister        rbacv1listers.RoleLister
+	roleBindingLister rbacv1listers.RoleBindingLister
+	rbacClient        rbacv1client.RbacV1Interface
+}
+
+// NewAddFinalizerController returns a AddFinalizerController
+func NewAddFinalizerController(
+	recorder events.Recorder,
+	roleInformer rbacv1informers.RoleInformer,
+	roleBindingInformer rbacv1informers.RoleBindingInformer,
+	rbacClient rbacv1client.RbacV1Interface,
+) factory.Controller {
+	controller := &AddFinalizerController{
+		roleLister:        roleInformer.Lister(),
+		roleBindingLister: roleBindingInformer.Lister(),
+		rbacClient:        rbacClient,
+	}
+
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			key, _ := cache.MetaNamespaceKeyFunc(obj)
+			return key
+		}, roleInformer.Informer(), roleBindingInformer.Informer()).
+		WithSync(controller.sync).ToController("AddFinalizerController", recorder)
+}
+
+// sync adds finalizer to role/rolebinding for work agent in the spoke cluster namespace. No matter which
+// resource trigger the sync, both role and rolebinding will be reconciled.
+func (m *AddFinalizerController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	key := controllerContext.QueueKey()
+
+	// ignore role/rolebinding whose name does not end with the expected suffix
+	if !strings.HasSuffix(key, keySuffix) {
+		return nil
+	}
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		// ignore role/rolebinding whose key is not in format: namespace/name
+		return nil
+	}
+	clusterName := strings.TrimSuffix(name, keySuffix)
```

Comment:
logic like this concerns me and it means that the linkage isn't really any better since the two repos are still coordinating